File: backend/apps/api/schemas/projects.py
from __future__ import annotations

from ninja import Schema


class ProjectRepositoryIn(Schema):
    repoSlug: str
    repositoryType: str = "git"
    repositoryProvider: str = "bitbucket"
    repoType: str = "bitbucket"
    repoKey: str = ""
    cloneUrl: str = ""
    defaultBranch: str = ""
    coverityProject: str = ""
    coverityStream: str = ""
    isRepresentative: bool = False


class ProjectBuildIn(Schema):
    buildName: str
    buildType: str = ""
    runtimeStack: str = ""
    buildId: str
    planKey: str
    repositorySlug: str


class BuildUnitProviderDetailsIn(Schema):
    planKey: str = ""
    buildId: str = ""
    jobPath: str = ""
    jobType: str = ""
    folderPath: str = ""
    pipelineKind: str = ""


class BuildUnitIn(Schema):
    externalKey: str
    displayName: str
    description: str = ""
    unitType: str = "build"
    repositorySlug: str = ""
    language: str = ""
    compiler: str = ""
    runtimeStack: str = ""
    lifecycleStatus: str = "active"
    isEnabled: bool = True
    providerDetails: BuildUnitProviderDetailsIn | None = None


class ProjectCreateIn(Schema):
    projectKey: str = ""
    name: str = ""
    description: str = ""
    ownerTeam: str = ""
    serviceType: str = ""
    ciProvider: str = "bamboo"
    status: str = "active"
    jiraProjectKey: str = ""
    bitbucketProjectKey: str = ""
    representativeRepoSlug: str = ""
    repositories: list[ProjectRepositoryIn] = []
    builds: list[ProjectBuildIn] = []
    buildUnits: list[BuildUnitIn] = []


class ProjectUpdateIn(Schema):
    name: str = ""
    description: str = ""
    ownerTeam: str = ""
    serviceType: str = ""
    ciProvider: str = "bamboo"
    status: str = "active"
    bitbucketProjectKey: str = ""
    representativeRepoSlug: str = ""
    repositories: list[ProjectRepositoryIn] = []
    builds: list[ProjectBuildIn] = []
    buildUnits: list[BuildUnitIn] = []

File: backend-rs/src/api/models/projects.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectListQuery {
    #[serde(default)]
    pub ci_provider: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ProjectCreateIn {
    pub jira_project_key: String,
    pub bitbucket_project_key: String,
    #[serde(default = "default_ci_provider")]
    pub ci_provider: String,
    pub display_name: String,
    pub description: Option<String>,
}

fn default_ci_provider() -> String {
    "bamboo".to_string()
}

#[derive(Debug, Deserialize)]
pub struct ProjectUpdateIn {
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub representative_repo_slug: Option<String>,
}
